pywhy_stats/kernel_utils.py

```diff
@@ -0,0 +1,281 @@
+
+    Returns
+    -------
+    data : ArrayLike of shape (n_features, n_features)
+        A symmetric centered correntropy matrix of the data.
+
+    Notes
+    -----
+    The estimator for the correntropy array is given by the formula
+    :math:`1 / N \\sum_{i=1}^N k(x_i, y_i) - 1 / N**2 \\sum_{i=1}^N \\sum_{j=1}^N k(x_i, y_j)`.
+    The first term is the estimate, and the second term is the bias, and together they form
+    an unbiased estimate.
+    """
+    n_samples, n_features = data.shape
+    corren_arr = np.zeros(shape=(n_features, n_features))
+
+    # compute kernel between each feature, which is now (n_features, n_features) array
+    for idx in range(n_features):
+        for jdx in range(idx + 1):
+            K = compute_kernel(
+                X=data[:, [idx]],
+                Y=data[:, [jdx]],
+                metric=metric,
+                centered=centered,
+                n_jobs=n_jobs,
+            )
+
+            # compute the bias due to finite-samples
+            bias = np.sum(K) / n_samples**2
+
+            # compute the sample centered correntropy
+            corren = (1.0 / n_samples) * np.trace(K) - bias
+
+            corren_arr[idx, jdx] = corren_arr[jdx, idx] = corren
+    return corren_arr
+
+
+def von_neumann_divergence(A: ArrayLike, B: ArrayLike) -> float:
+    """Compute Von Neumann divergence between two PSD matrices.
+
+    Parameters
+    ----------
+    A : ArrayLike of shape (n_features, n_features)
+        The first PSD matrix.
+    B : ArrayLike of shape (n_features, n_features)
+        The second PSD matrix.
+
+    Returns
+    -------
+    div : float
+        The divergence value.
+
+    Notes
+    -----
+    The Von Neumann divergence, or what is known as the Bregman divergence in
+    :footcite:`Yu2020Bregman` is computed as follows with
+    :math:`D(A || B) = Tr(A (log(A) - log(B)) - A + B)`.
+    """
+    div = np.trace(A.dot(logm(A) - logm(B)) - A + B)
+    return div
+
+
+def _preprocess_kernel_data(
+    X: ArrayLike,
+    Y: ArrayLike,
+    Z: Optional[ArrayLike] = None,
+    normalize_data: bool = True,
+):
+    """Preprocess the data for kernel methods.
+
+    Parameters
+    ----------
+    X : ArrayLike of shape (n_samples_X, n_features_X)
+        The X array.
+    Y : ArrayLike of shape (n_samples_Y, n_features_Y)
+        The Y array.
+    Z : ArrayLike of shape (n_samples_Z, n_features_Z), optional
+        The Z array, by default None.
+    normalize_data : bool
+        Whether to standard-normalize the data or not.
+    """
+    X = np.array(X)
+    Y = np.array(Y)
+    if Z is not None:
+        Z = np.array(Z)
+
+    if X.ndim == 1:
+        X = X.reshape(-1, 1)
+    if Y.ndim == 1:
+        Y = Y.reshape(-1, 1)
+    if Z is not None and Z.ndim == 1:
+        Z = Z.reshape(-1, 1)
+
+    # handle strings as categorical data automatically
+    if X.dtype.type in (np.str_, np.object_, np.unicode_):
+        enc = LabelEncoder()
```

Comment:
Do you have a strong preference for leaving the `np.str_` support in? Idr why, but I think I was running into a bunch of errors on the string unit-test; I'll need to look at how to revert this. In general, I find working internally assuming `ArrayLike[numbers.Real]` type easier to maintain, so we don' have separate code paths if something is a string. 

I suppose the main thing is being able to auto-assume something is categorical, with the caveat that the strings are correct. However, is the benefit worth it? 

Let me investigate in the separate PR to see if I can demonstrate what I mean.

Reply:
In DoWhy-GCM, we have the convention that categorical data are strings. So, if I want to use this method with string data and do not define a particular kernel (which majority of the users won't), it would not work. The easiest convention is to define all string values as categorical. I actually simply checked before if the first element in a column value is a `str`, then we don't need all this `np.str_` types. 

An alternative (but technically less accurate) is to introduce something like `encode_strings: bool = False`, default to RBF kernels and then throw an error with string data if `encode_strings` is set to False. But again, applying the delta kernel would be more accurate.